applications/IgaApplication/custom_conditions/support_penalty_condition.cpp

```diff
@@ -0,0 +1,176 @@
+
+            //FOR DISPLACEMENTS
+            Matrix H = ZeroMatrix(3, mat_size);
+            for (IndexType i = 0; i < number_of_nodes; i++)
+            {
+                IndexType index = 3 * i;
+                H(0, index)     = N(point_number, i);
+                H(1, index + 1) = N(point_number, i);
+                H(2, index + 2) = N(point_number, i);
+            }
+
+            // Differential area
+            const double integration_weight = integration_points[point_number].Weight();
+            const double determinat_jacobian = determinat_jacobian_vector_initial[point_number];
+
+            // Assembly
+            if (CalculateStiffnessMatrixFlag) {
+                noalias(rLeftHandSideMatrix) += prod(trans(H), H)
+                    * penalty * integration_weight * determinat_jacobian;
+            }
+            if (CalculateResidualVectorFlag) {
+
+                const array_1d<double, 3>& displacement = Has(DISPLACEMENT)
+                    ? this->GetValue(DISPLACEMENT)
+                    : ZeroVector(3);
+
+                Vector u(mat_size);
+                for (IndexType i = 0; i < number_of_nodes; i++)
+                {
+                    const array_1d<double, 3> disp = r_geometry[i].FastGetSolutionStepValue(DISPLACEMENT);
+                    IndexType index = 3 * i;
+                    u[index] = (disp[0] - displacement[0]);
+                    u[index + 1] = (disp[1] - displacement[1]);
+                    u[index + 2] = (disp[2] - displacement[2]);
+                }
+
+                noalias(rRightHandSideVector) -= prod(prod(trans(H), H), u)
+                    * penalty * integration_weight * determinat_jacobian;
+            }
+        }
+        KRATOS_CATCH("")
+    }
+
+    void SupportPenaltyCondition::DeterminantOfJacobianInitial(
+        const GeometryType& rGeometry,
+        Vector& rDeterminantOfJacobian)
+    {
+        const IndexType nb_integration_points = rGeometry.IntegrationPointsNumber();
+        if (rDeterminantOfJacobian.size() != nb_integration_points) {
+            rDeterminantOfJacobian.resize(nb_integration_points, false);
+        }
+
+        const SizeType working_space_dimension = rGeometry.WorkingSpaceDimension();
+        const SizeType local_space_dimension = rGeometry.LocalSpaceDimension();
+        const SizeType nb_nodes = rGeometry.PointsNumber();
+
+        Matrix J = ZeroMatrix(working_space_dimension, local_space_dimension);
+        for (IndexType pnt = 0; pnt < nb_integration_points; pnt++)
+        {
+            const Matrix& r_DN_De = rGeometry.ShapeFunctionsLocalGradients()[pnt];
+            J.clear();
+            for (IndexType i = 0; i < nb_nodes; ++i) {
+                const array_1d<double, 3>& r_coordinates = rGeometry[i].GetInitialPosition();
+                for (IndexType k = 0; k < working_space_dimension; ++k) {
+                    const double value = r_coordinates[k];

```

Comment:
isn't better to declare value outside the loop as non `const` and overwritting it in the loop, instead of redeclaring a `const` variable in each loop

Reply:
I don't think storing it does makes anything better anyways...